packages/api-server/src/methods/modules/poly.ts

```diff
@@ -142,3 +168,113 @@ export class Poly {
     }
   }
 }
+
+async function saveAddressMapping(
+  query: Query,
+  rpc: GodwokenClient,
+  txWithAddressMapping:
+    | L2TransactionWithAddressMapping
+    | RawL2TransactionWithAddressMapping
+) {
+  console.log(JSON.stringify(txWithAddressMapping, null, 2));
+
+  if (
+    txWithAddressMapping.addresses.length === "0x0" ||
+    txWithAddressMapping.addresses.data.length === 0
+  ) {
+    console.log(`empty addressMapping, abort saving.`);
+    return;
+  }
+
+  if (txWithAddressMapping.extra === EMPTY_ABI_ITEM_SERIALIZE_STR) {
+    console.log(`addressMapping without abiItem, abort saving.`);
+    return;
+  }
+
+  let rawTx;
+  if ("raw_tx" in txWithAddressMapping) {
+    rawTx = txWithAddressMapping.raw_tx;
+  } else {
+    rawTx = txWithAddressMapping.tx.raw;
+  }
+  const ethTxData = decodeArgs(rawTx.args).data;
+  const abiItemStr = txWithAddressMapping.extra;
+  const abiItem: AbiItem = deserializeAbiItem(abiItemStr);
+  const addressesFromEthTxData = getAddressesFromInputDataByAbi(
+    ethTxData,
+    abiItem
+  );
+  if (addressesFromEthTxData.length === 0) {
+    console.log(
+      `eth tx data ${ethTxData} contains no valid address, abort saving.`
+    );
+    return;
+  }
+
+  txWithAddressMapping.addresses.data.forEach(async (item) => {
```

Comment:
sequence: `for (const item of txWithAddressMapping.addresses.data)`
parallel: `await Promise.all(txWithAddressMapping.addresses.data.map(async (item) => {))`

Reply:
updated with parallel.